### PR TITLE
[ENG-3110] NotableEmailDomain (and a bit of domain-based trust)

### PR DESCRIPTION
**Pull request by aaxelb**
## Purpose
Avoid effort wasted dealing with automated spam-check false positives.

## Changes
* rename `BlacklistedEmailDomain` to `NotableEmailDomain`
* add `NotableEmailDomain.note` field with two possible values:
  * `EXCLUDE_FROM_ACCOUNT_CREATION`: existing behavior of `BlacklistedEmailDomain`, prevent creating accounts with emails from this domain
  * `ASSUME_NONSPAM_UNTIL_REPORTED`: (aka "trusted") if a user has an email from this domain, skip spam checks on content they post
* allow bulk-adding NotableEmailDomains in the admin interface

## QA Notes
gonna dev-test

Please make verification statements inspired by your code and what your code touches.
- Verify spam checks run for users without a trusted email domain
- Verify spam checks don't run for users with a trusted email domain

What are the areas of risk?
- touched logic for existing "exclude from account creation" domains -- should still work as before

Any concerns/considerations/questions that development raised?

## Documentation
- internal support docs for osf-admin:
  - meaning/effects of NotableEmailDomain's two notes
  - how to view/search/delete/edit/add/bulk-add NotableEmailDomains

## Side Effects
n/a

## Ticket
[ENG-3110]

[ENG-3110]: https://openscience.atlassian.net/browse/ENG-3110?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by aaxelb** [comment]
@Johnetordoff hmm my intent was "known to us", since the table now contains both trusted and untrusted domains (but not the millions of domains we have no opinion on) -- am open to suggestions that better captures that

since we already had `BlacklistedEmailDomain`, this seemed like a fine opportunity to get rid of a problematic name and build on what already exists, instead of adding another near-identical table

**Comment by Johnetordoff** [comment]
I'm just trying to think about it a little more practically, if we're talking about this in a meeting with devs who are unfamiliar with it they are not going to know if these domains are banned or verified, I'd typically think if we "know" an email address we must trust it.

**Comment by aaxelb** [comment]
i can see how "known" might carry a friendly implication -- how about `RecognizedEmailDomain`? or could rephrase to match the two model fields -- `EmailDomainTrustLevel`, `EmailDomainTrustworthiness`?

**Comment by jwalz** [comment]
> i can see how "known" might carry a friendly implication -- how about `RecognizedEmailDomain`? or could rephrase to match the two model fields -- `EmailDomainTrustLevel`, `EmailDomainTrustworthiness`?

Definitely agree that "KnownEmailDomain" sounds like an allow-list if you're not looking at the implementation. I would suggest something like "ActionableEmailDomain"? And, for ease of use, manager-level methods to get all blocked or trusted domains?

**Comment by jwalz** [comment]
Also, one of the reasons tests are dying is because https://github.com/CenterForOpenScience/osf.io/pull/9741 got merged and your migration is now in conflict. Sorry for not warning you as I did the Bulk Upload folks

**Comment by aaxelb** [comment]
hmm i definitely would not understand what "actionable email domain" means without digging further (might argue the domain alone isn't actionable -- also need an email address in order to act)

an instance of this model is an email domain that we've annotated with one bit of info...
`AnnotatedEmailDomain`?
`NotedEmailDomain`?
`FlaggedEmailDomain`?
`TaggedEmailDomain`?

File: tests/test_spam_mixin.py [hunk 1]
```diff
@@ -117,45 +135,115 @@ def test_remove_flag(self):
...
+    def test_confirm_ham(self, spammable_thing):
+        spammable_thing.confirm_ham(save=True)
+        assert spammable_thing.is_ham
+
+    def test_confirm_spam(self, spammable_thing):
+        spammable_thing.confirm_spam(save=True)
+        assert spammable_thing.is_spam
+
+    @pytest.mark.parametrize('assume_ham', (True, False))
+    def test_spam_status_properties(self, spammable_thing, assume_ham):
```
[the remark this answers is not in the document]
Comment (aaxelb) [reply tests/test_spam_mixin.py]: yeah that makes sense -- parametrized by SpamStatus, but kept the complication of testing `assume_ham` for all states (make sure it doesn't override an explicit spam flag or confirmation)

**Review by jwalz — APPROVED** [review]

**Pull request merged**